File: app/src/main/java/com/numero/colorpalette/view/ColorStateView.java
package com.numero.colorpalette.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.numero.colorpalette.R;

public class ColorStateView extends FrameLayout {

    private View parentLayout;
    private TextView titleTextView;
    private TextView colorNameTextView;

    public ColorStateView(@NonNull Context context) {
        this(context, null);
    }

    public ColorStateView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ColorStateView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        View view = LayoutInflater.from(context).inflate(R.layout.view_color_state, this, false);
        initViews(view);
        addView(view);

        final TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.ColorStateView);
        String title = a.getString(R.styleable.ColorStateView_title);
        setTitle(title);
        a.recycle();
    }

    private void initViews(@NonNull View view) {
        parentLayout = view.findViewById(R.id.parent_layout);

        titleTextView = view.findViewById(R.id.title_text_view);
        colorNameTextView = view.findViewById(R.id.color_name_text);
    }

    public void setTitle(@Nullable String titleString) {
        titleTextView.setText(titleString);
    }

    public void setColorName(@Nullable String colorName) {
        colorNameTextView.setText(colorName);
    }

    @Override
    public void setOnClickListener(@Nullable OnClickListener l) {
        parentLayout.setOnClickListener(l);
    }
}

File: app/src/main/java/com/numero/colorpalette/view/ColorStateView.kt
package com.numero.colorpalette.view

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout

import com.numero.colorpalette.R
import com.numero.colorpalette.util.ColorUtil

import io.reactivex.Observable
import kotlinx.android.synthetic.main.view_color_state.view.*

class ColorStateView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : FrameLayout(context, attrs, defStyleAttr) {

    var titleText: String? = null
        set(value) {
            titleTextView.text = value
        }

    var colorCodeText: String? = null
        set(value) {
            colorCodeTextView.text = value
            setColorCode(value)
        }

    init {
        View.inflate(context, R.layout.view_color_state, this)

        context.obtainStyledAttributes(attrs, R.styleable.ColorStateView).apply {
            titleText = getString(R.styleable.ColorStateView_title)
            colorCodeText = getString(R.styleable.ColorStateView_colorCode)
            recycle()
        }
    }

    override fun setOnClickListener(l: View.OnClickListener?) {
        parentLayout.setOnClickListener(l)
    }

    private fun setColorCode(code: String?) {
        if (code.isNullOrEmpty()) {
            visibility = View.GONE
            return
        }
        visibility = View.VISIBLE
        Observable.just(code)
                .map { Color.parseColor(it) }
                .doOnNext {
                    setBackgroundColor(it)
                }
                .map { ColorUtil.getTextColor(it) }
                .subscribe({
                    titleTextView.setTextColor(it)
                    colorCodeTextView.setTextColor(it)
                }, {
                    visibility = GONE
                })
    }
}
